system/datatypes/mailinglist.php

```diff
@@ -0,0 +1,959 @@
+		/** @var HyphaDomElement $mailings */
+		$mailings = $this->getDoc()->getOrCreate('mailings');
+		$mailings->append($mailing);
+		$this->xml->saveAndUnlock();
+	}
+
+	/**
+	 * @return bool
+	 */
+	private function hasSender() {
+		$senderData = $this->getSenderData();
+
+		return '' != $senderData['name'] && '' != $senderData['email'];
+	}
+
+	/**
+	 * @return array
+	 */
+	private function getSenderData() {
+		$senderData = [
+			'name' => $this->getDoc()->getAttribute('sender-name'),
+			'email' => $this->getDoc()->getAttribute('sender-email'),
+		];
+		if ('' == $senderData['name']) {
+			$senderData['name'] = hypha_getTitle();
+		}
+		if ('' == $senderData['email']) {
+			$senderData['email'] = hypha_getEmail();
+		}
+
+		return $senderData;
+	}
+
+	/**
+	 * @param int $length
+	 *
+	 * @return string
+	 */
+	private function constructCode($length = 16) {
+		try {
+			if (function_exists('random_bytes')) {
+				return bin2hex(random_bytes($length));
+			}
+			if (function_exists('mcrypt_create_iv')) {
+				return bin2hex(mcrypt_create_iv($length, MCRYPT_DEV_URANDOM));
+			}
+		} catch (\Exception $e) {
+		}
+
+		return bin2hex(openssl_random_pseudo_bytes($length));
+	}
+
+	/**
+	 * @todo [LRM]: move so it can be used throughout Hypha
+	 * @param DOMElement $elem
+	 * @param array $data
+	 *
+	 * @return WymHTMLForm
+	 */
+	private function createForm(DOMElement $elem, array $data = []) {
+		$form = new WymHTMLForm($elem);
+		$form->setData($data);
+
+		return $form;
+	}
+
+	/**
+	 * @todo [LRM]: move so it can be used throughout Hypha
+	 * @param string $label
+	 * @param null|string $path
+	 * @param null|string $command
+	 * @param null|string $argument
+	 *
+	 * @return string
+	 */
+	private function makeActionButton($label, $path = null, $command = null, $argument = null) {
+		$path = $this->language . '/' . $this->pagename . ($path ? '/' . $path : '');
+		$_action = makeAction($path, ($command ? $command : ''), ($argument ? $argument : ''));
+
+		return makeButton(__($label), $_action);
+	}
+
+	/**
+	 * @todo [LRM]: move so it can be used throughout Hypha
+	 * @param string $selector
+	 *
+	 * @return NodeList
+	 */
+	private function findBySelector($selector) {
+		return $this->html->find($selector);
+	}
+
+	/**
+	 * @todo [LRM]: move so it can be used throughout Hypha
+	 * @param string $path
+	 * @param null|string $language
+	 *
+	 * @return string
+	 */
+	private function constructFullPath($path, $language = null) {
+		global $hyphaUrl;
+		$language = null == $language ? $this->language : $language;
+		$path = '' == $path ? '' : '/' . $path;
+
+		return $hyphaUrl . $language . $path;
+	}
+
+	/**
+	 * @todo [LRM]: move so it can be used throughout Hypha
+	 * @param null|string $command
+	 *
+	 * @return bool
+	 */
+	private function isPosted($command = null) {
+		return 'POST' == $_SERVER['REQUEST_METHOD'] && (null == $command || $command == $_POST['command']);
+	}
+
+	/**
+	 * @todo [LRM]: move so it can be used throughout Hypha
```

Comment:
You should probably make this "and/or integrate this with `sendMail() and/or let it make use of `sendMail()`" (or perhaps just make it use `sendMail()` now already?)